File: 2024/11/main.py
import sys

def process_stone(stone: int) -> list[int]:
    size = len(str(stone))
    if stone == 0:
        return [1]
    elif size % 2 == 0:
        mid = int(size / 2)
        return [
            int(str(stone)[:mid]),
            int(str(stone)[mid:]),
        ]
    else:
        return [stone * 2024]

def main():
    stones = [
        int(stone)
        for stone in sys.stdin.readline().strip().split()
    ]

    for _ in range(25):
        stones = [
            new_stone
            for stone in stones
            for new_stone in process_stone(stone)
        ]

    print(f"[Part one] {len(stones)} stones")

if __name__ == "__main__":
    main()

File: 2024/11/src/main.rs
use std::str::FromStr;
use std::io::BufRead;
use std::collections::HashMap;

fn process_stones(iteration: i32, stone: i64, cache: &mut HashMap<(i32, i64), i64>) -> i64 {
    if iteration == 0 {
        return 1;
    }

    if let Some(count) = cache.get(&(iteration, stone)) {
        return *count;
    }

    let mut result = 0;
    if stone == 0 {
        result = process_stones(iteration - 1, 1, cache);
    } else {
        let size = stone.ilog10() + 1;
        if size % 2 == 0 {
            let divider = 10_i64.pow(size / 2);
            result += process_stones(iteration - 1, stone % divider, cache);
            result += process_stones(iteration - 1, stone / divider, cache);
        } else {
            result = process_stones(iteration - 1, stone * 2024, cache);
        }
    }

    cache.insert((iteration, stone), result);
    return result
}

fn main() {
    let stdin = std::io::stdin().lock();
    let line = stdin.lines().next().unwrap().unwrap();
    let mut cache: HashMap<(i32, i64), i64> = HashMap::new();

    let count: i64 = line.split(" ")
        .map(|stone| (25, i64::from_str(stone).unwrap()))
        .map(|(iteration, stone)| process_stones(iteration, stone, &mut cache))
        .sum();
    println!("[Part one] {} stones", count);

    let count: i64 = line.split(" ")
        .map(|stone| (75, i64::from_str(stone).unwrap()))
        .map(|(iteration, stone)| process_stones(iteration, stone, &mut cache))
        .sum();
    println!("[Part two] {} stones", count);
}
